Fix submit for review button

oandregal opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/10475

This PR takes the viewport size into account when it comes to decide whether we show the button or toggle logic. Here's how this PR modifies the logic (changes in bold):

Logic | Before | After 
--- | --- | ---
Post status is `publish` or `private` | Button | Button
Post status is `future` and the (scheduled) date is after now | Button | Button
Post status is `pending`, cannot be published and the viewport is < medium  | **Button** | **Toggle**
Post status is `pending`, cannot be published and the viewport >= medium | Button | Button
Publish sidebar is enabled | Toggle | Toggle
Publish sidebar is disabled and viewport is < medium | **Button** | **Toggle**
Publish sidebar is disabled and viewport is >= medium | Button | Button

### TODO

- [x] Show the "Publish..." button below medium viewports overriding the pre-publish checks option if necessary.
- [x] Hide pre-publish checks option on the general settings modal below medium viewports.
- [x] Update/Add e2e tests.

oandregal committed:
Change check order

oandregal committed:
Add isPublished check to header

oandregal committed:
Add scheduling check to header

oandregal committed:
Add publish action check to header

oandregal committed:
Remove unnecessary code

oandregal committed:
Force the Publish button on small viewports

jasmussen commented:
Hmm am I doing something wrong?

I have unchecked prepublish checks, but they're still not present on mobile:

<img width="447" alt="screen shot 2018-10-23 at 12 55 14" src="https://user-images.githubusercontent.com/1204802/47355925-fd82b000-d6c2-11e8-9614-b0397351a65d.png">

oandregal committed:
Hide for viewports < medium

Some languages such as Deustch have longer text that need acomodation.
We are also using medium to hide the writing modes,
so it makes sense to follow that.

nosolosw commented:
> Hmm am I doing something wrong?

Testing before this is ready! :smile:

_I'm still working on this. I should have waited before pinging you as a reviewer._

jasmussen commented:
No worries. I'll keep an eye on this!

oandregal committed:
Rename variable for clarity

oandregal committed:
Override the isPublishSidebarEnabled option in < medium viewports

We want to show the "Publish..." button on < medium viewports,
no matter what the isPublishSidebarEnabled state is.

oandregal committed:
Hide pre-publish check options in < medium viewports

nosolosw commented:
@jasmussen @WordPress/gutenberg-core this is ready for review. I still need to get ready the e2e tests, but wanted to give folks as much time as possible for review. Need to be away a couple of hours but will come back to this by the end of the day.

oandregal committed:
Add e2e test to check buttons

nosolosw commented:
Added some basic e2e tests in c7b89a2.

oandregal committed:
Update snapshot

noisysocks commented on test/e2e/specs/publishing.test.js:
```diff
@@ -76,4 +77,30 @@ describe( 'Publishing', () => {
+				await setViewport( 'small' );
+			} );
+
+			afterEach( async () => {
```
Why have the `beforeEach` and `afterEach` callbacks if we only have one test?

nosolosw replied:
Because I was hoping to add more tests, but I've failed to. I would still leave them as they are, in case I can revisit before this is merged.

noisysocks commented on packages/edit-post/src/components/options-modal/options.js:
```diff
@@ -51,7 +52,8 @@ export const EnablePublishSidebarOption = compose(
 		};
-	} )
+	} ),
+	ifViewportMatches( 'medium' ),
```
It would be helpful to have a little comment here that explains that small viewports **always** have the publish sidebar.

nosolosw replied:
c567dee

noisysocks commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -29,10 +35,22 @@ function Header( {
+		isPublished ||
+		( isScheduled && isBeingScheduled ) ||
+		( isPending && ! hasPublishAction && ! isSmallViewport )
+	);
```
I'm having a hard time understanding this expression. Could we refactor it as a function which uses early `return`s? Exporting said function would then let us write unit tests to test the logic.

noisysocks reviewed (commented):
This works in my testing and is looking really good! Great work 💯📈🎉

I left a few notes on how I think we could make the code a little easier to understand.

oandregal committed:
Add comment

oandregal committed:
Extract PostPublishButtonOrToggle component

oandregal committed:
More refactoring

oandregal committed:
Test that button is rendered when post is published

oandregal committed:
Tests for publish sidebar not enabled

oandregal committed:
Tests for post scheduled

oandregal committed:
Tests for post pending that cannot be published

oandregal committed:
Better describe tests

oandregal committed:
Better describe test

oandregal committed:
Refactor conditions to show toggle or button

oandregal committed:
Add back the isPending and cannot be published for > small viewports check

oandregal committed:
Add test for is pending status

oandregal committed:
Rename variables and add tests

nosolosw commented:
@noisysocks done a refactoring to clarify the button or toggle logic, as you suggested. Also added a bunch of unit tests. Everything is much clearer now! Added a table to the PR description with the changes and some [code comments](https://github.com/WordPress/gutenberg/pull/10941/files#diff-3bdc256a8b81e9a0d735c54f77d35328R43) to add further context to our futureselves.

oandregal committed:
Bring back a property deleted by accident

noisysocks committed:
Format post-publish-button-or-toggle.js using prettier

noisysocks reviewed (approved):
Yes! Love it! This is a lot easier to follow. Thanks also for documenting the logic both in the code and in the PR. Gold star for you ⭐️ 

There were some minor indentation issues in `post-publish-button-or-toggle.js`. I fixed them up by running Prettier on the file in f4aed3d. Hope you don't mind!

noisysocks commented on packages/editor/src/components/post-publish-panel/toggle.js:
```diff
@@ -64,23 +44,15 @@ export default compose( [
 			isPublished: isCurrentPostPublished(),
-			isScheduled: isCurrentPostScheduled(),
 			isBeingScheduled: isEditedPostBeingScheduled(),
-			postType: getCurrentPostType(),
```
I've noticed we're selecting stuff that we don't need all over the place. I wonder if we could somehow write a lint rule to catch them all...

nosolosw commented:
Thanks for `f4aed3d`, Robert!
